enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+	// For route based setting of headers
+	SetHTTPResponseHeaders              []SetHTTPHeader
+	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
+}
+```
+```go
+// SetHTTPHeader specifies an HTTP header that should be set.
+type SetHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+
+	// Value specifies the header value.
+	Value string
+
+	// Action specifies the action to be performed.
+	Action string
+}
+
+type DeleteHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
+on either `spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` which will have the desired
+sequence.
+
+C) We will be creating an immutable/harcoded deny list of HTTP Response headers in the [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141) 
+which won't be allowed to be set/delete. Eg - HSTS.
+
+### Risks and Mitigations
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+- Would this enhacement cover setting request headers ?
+
+  Answer: As per [epic](https://issues.redhat.com/browse/RFE-464) the requirement for the customers seems to be for HTTP response headers only.
+  We need confirmation from the PM regarding this.
+
+- Would we be able to set load balancing alogorithm based on headers ?
+
+  Answer: No, this out of scope for this enhancement so it will be considered as a non-goal.
+
+- Would we be able to delete or append headers  ?
+
+  Answer: Yes, this enhancement will cover deletion and but not appending of HTTP Response Headers as per the [epic](https://issues.redhat.com/browse/RFE-464). 
+
+- Would some headers be blocked for setting and deleting ?
+
+  Answer: Yes, this enhancement won't allow the user to set or delete HSTS headers. We already a way to set this which is via route 
+  annotation. Providing one more way to add this can cause issues.
+
+- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression)?
+  Answer: No, this is out of scope for this enhancement.
+
+- Would we be able to [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression) ?
+  Answer:  No, this is out of scope for this enhancement.
+
```

Comment:
One of the goals of this EP:
>  Define an immutable/hard-coded deny list of headers which are not allowed to be set or deleted.

This would forbid certain headers from being set/delete via the new API. However the use case you described is a little more complicated. `Cache-Control: private` is set by HAProxy's `cookie nocache` command which is needed for the [right sticky session implementation](https://www.haproxy.com/blog/enable-sticky-sessions-in-haproxy/#implement-sticky-sessions-with-a-cookie). The OpenShift IngressController follows this recommendation. So, since `Cache-Control: private` is set outside of this API, the previously mentioned deny list won't forbid it.

Also, this API will give you the possibility to remove this header and add it with the values you want unless we decide to forbid to do for `Cache-Control` header via the deny list (will be decided later).

Reply:
> Also, this API will give you the possibility to remove this header and add it with the values you want unless we decide to forbid to do so for the `Cache-Control` header via the deny list (will be decided later).

@alebedev87, @miheer, is one of you working on reaching that decision?

Reply:
It would be better to resolve the matter before we move on from the EP to the implementation.  At the very least, the EP should note that the `cache-control` header may be on the deny list.  

By the way, are you saying that we might need two deny lists: one for the IngressController API and one for the Route API?

Reply:
> By the way, are you saying that we might need two deny lists: one for the IngressController API and one for the Route API?

Not really. I just meant that it's not quite clear how HAProxy config execution will behave. Example:
```
backend be_secure:openshift-console:console
  http-response del-header Cache-Control
  cookie 1e2670d92730b515ce3a1bb65da45062 nocache ...
```
Knowing that `cookie` command will set `Cache-Control` header, will we still have it at end of the backend processing?

However we [know](http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#4.2-http-response) that the `frontend` section will be processed after `backend`, so adding `http-response del-header Cache-Control` will remove the header.

Reply:
@Miciah black listing  of headers by users is a not a part of this epic. It will increase the scope.

Reply:
if we want a deny list it will be like a new feature. We will need to revise the API as well.

Reply:
For now as you rightly mentioned we can add cache-control in library-go code.

Reply:
Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests. Only the dynamic requests need to stick to the same server (session sticky). 

Most applications will set something like this for static assets:
```
cache-control: max-age=7200, must-revalidate, public
```

But the OpenShift router (HAProxy) will add `cache-control: private` which results in both headers in the response:

```
cache-control: private 
cache-control: max-age=7200, must-revalidate, public
```

This leads to issues with caching: CDNs like Cloudflare will not cache your files. 

The best approach for users (developers) here would be to only add the `cache-control: private` if there is no `cache-control: public` already present in the response.

Reply:
> Many applications serve mixed content: static assets (images, js, css) and dynamic content (html, json). They usually want to cache the static assets on the client and do not care about session stickiness for those requests.

A common technique is to put static assets on a different virtual host.  In OpenShift terms, that means using a distinct route for static assets.  This route can be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting `cache-control`.  Would that solution suffice?

Reply:
Don't forget that we also hard-code or explicitly configure some headers within the HAProxy template, e.g. https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L337-L348   and https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L537-L541, and some others.

Reply:
Can you please check https://github.com/openshift/router/pull/438#discussion_r1170637633 ?